tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Source.Tests/DiagnosticAnalysisRunnerTests.cs

```diff
@@ -0,0 +1,116 @@
+        private readonly Fixture _fixture;
+
+        public DiagnosticAnalysisRunnerTests()
+        {
+            _fixture = new Fixture();
+        }
+
+        [Fact]
+        public async Task DiagnosticsWithEmptyProjectTest()
+        {
+            // Arrange
+            using var mock = AutoMock.GetLoose();
+            var loggerMock = new Mock<ILogger<RoslynDiagnosticProvider>>();
+
+            var project = CreateProject(mock);
+            project.Setup(p => p.GetRoslynProject()).Returns(() =>
+            {
+                var ws = new AdhocWorkspace();
+                var name = Path.GetFileNameWithoutExtension("TestProject.csproj")!;
+                return ws.AddProject(ProjectInfo.Create(ProjectId.CreateNewId(), VersionStamp.Default, name, name, "C#", filePath: "TestProject.csproj"));
+            });
+
+            var context = mock.Mock<IUpgradeContext>();
+            context.Setup(c => c.Projects).Returns(new[] { project.Object });
+            var diagnosticAnalysisRunner = new RoslynDiagnosticProvider(GetAnalyzers(false), Array.Empty<AdditionalText>(), loggerMock.Object);
+
+            // Act
+            var diagnostics = await diagnosticAnalysisRunner.GetDiagnosticsAsync(project.Object, CancellationToken.None).ConfigureAwait(false);
+
+            // Assert
+            Assert.Empty(diagnostics);
+        }
+
+        [Theory]
+        [InlineData(false, 0, null)]
+        [InlineData(true, 1, "Test1")]
+        public async Task DiagnosticsWithProjectFilesTest(bool includeLocation, int expectedDiagnosticCount, string? expectedDiagnosticId)
+        {
+            // Arrange
+            using var mock = AutoMock.GetLoose();
+            var loggerMock = new Mock<ILogger<RoslynDiagnosticProvider>>();
+
+            var project = CreateProject(mock);
+            project.Setup(p => p.GetRoslynProject()).Returns(() =>
+            {
+                var ws = new AdhocWorkspace();
+                var name = Path.GetFileNameWithoutExtension("TestProject.csproj")!;
+                var projectInfo = ProjectInfo.Create(ProjectId.CreateNewId(), VersionStamp.Default, name, name, "C#", filePath: "TestProject.csproj");
+                var documents = new List<DocumentInfo>() { DocumentInfo.Create(DocumentId.CreateNewId(projectInfo.Id), "Foo.cs") };
+                var proj = ws.AddProject(projectInfo.WithDocuments(documents));
+                return proj;
+            });
+
+            var context = mock.Mock<IUpgradeContext>();
+            context.Setup(c => c.Projects).Returns(new[] { project.Object });
+            var diagnosticAnalysisRunner = new RoslynDiagnosticProvider(GetAnalyzers(includeLocation), Array.Empty<AdditionalText>(), loggerMock.Object);
+
+            // Act
+            var diagnostics = await diagnosticAnalysisRunner.GetDiagnosticsAsync(project.Object, CancellationToken.None).ConfigureAwait(false);
+
+            // Assert
+            Assert.True(diagnostics.Count() == expectedDiagnosticCount);
```

Comment:
Instead of this pattern, if you do a collection, it's better to do:

```cs
Assert.Collection(diagnostics,
  r => Assert....,
  .... [for each item]);
```

Reply:
One of the testdata is to check for empty diagnostics, in that case the Assert.Collection was failing which is why I had to put it under `if(diagnostics.Any())`.

Reply:
Yup, when `expectedDiagnosticCount == 0`. It's ok as is, but the collection ends up being a bit better for readability and understanding what went wrong when something fails

Reply:
Agreed, `Assert.Collection` provides better readability than the plain old `Assert.True / False`